src/ln/functional_tests.rs

```diff
@@ -6602,3 +6602,308 @@ fn test_onion_failure() {
 		msg.onion_routing_packet = onion_packet;
 	}, ||{}, true, Some(21), None);
 }
+
+#[test]
+fn test_update_add_htlc_bolt2() {
+	use util::rng;
+	use std::sync::atomic::Ordering;
+	use super::channelmanager::HTLCSource;
+	use super::channel::ChannelError;
+
+	let secp_ctx = Secp256k1::new();
+
+	// BOLT 2 Requirements for Sender
+	// BOLT 2 Requirement: MUST NOT offer amount_msat it cannot pay for in the remote commitment transaction at the current feerate_per_kw (see "Updating Fees") while maintaining its channel reserve.
+	//TODO: I don't believe this is explicitly enforced when sending an HTLC but as the Fee aspect of the BOLT specs is in flux leaving this as a TODO.
+
+	// BOLT2 Requirement: MUST offer amount_msat greater than 0.
+	// BOLT2 Requirement: MUST NOT offer amount_msat below the receiving node's htlc_minimum_msat (same validation check catches both of these)
+	let mut nodes = create_network(2);
+	let chan = create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 100000, 95000000);
+	let route = nodes[0].router.get_route(&nodes[1].node.get_our_node_id(), None, &[], 100000, TEST_FINAL_CLTV).unwrap();
+	let (_, our_payment_hash) = get_payment_preimage_hash!(nodes[0]);
+
+	let session_priv = SecretKey::from_slice(&secp_ctx, &{
+		let mut session_key = [0; 32];
+		rng::fill_bytes(&mut session_key);
+		session_key
+	}).expect("RNG is bad!");
+
+	let cur_height = nodes[0].node.latest_block_height.load(Ordering::Acquire) as u32 + 1;
+	let onion_keys = onion_utils::construct_onion_keys(&secp_ctx, &route, &session_priv).unwrap();
+	let (onion_payloads, _htlc_msat, _htlc_cltv) = onion_utils::build_onion_payloads(&route, cur_height).unwrap();
+	let onion_packet = onion_utils::construct_onion_packet(onion_payloads, onion_keys, &our_payment_hash);
+
+	let err = nodes[0].node.channel_state.lock().unwrap().by_id.get_mut(&chan.2).unwrap().send_htlc(0, our_payment_hash, TEST_FINAL_CLTV, HTLCSource::OutboundRoute {
```

Comment:
The reason is that from what I can see when using send_payment(...) the payment parameters are all set up when you call get_route(...) and most of these test cases cause various failures when trying to construct the route. I was trying to test all the validation that is done when adding the HTLC to be sent rather than constructing the route so to get into the middle of that process I constructed a valid route and then implemented the essential stuff that is happening inside nodes[0].node.send_payment(...) in the test and changed the payment parameters.

Reply:
Oh, I think we're out of sync on which send_payment. I dont mean the functional_tests::send_payment function that steps through everything, I meant the ChannelManager::send_payment, which calls Channel::send_htlc pretty directly and takes the Route as a parameter.

Reply:
Ok, I figured out how to effect these tests without skipping that layer. I was originally trying to check for the correct error types which was tricky because the various error handling macro's changed the types into more generic versions are they were handled but by just looking for the error message strings it is possible to do it as you suggest.